tests/org.dataflowanalysis.analysis.tests/src/org/dataflowanalysis/analysis/tests/dfd/MicroSecEndTest.java

```diff
@@ -0,0 +1,296 @@
+    // on the same host. Log data from microservices should not be send to the central logging system directly, but
+    // collected by the logging agent, written to a local file, and eventually send to the central system by it.
+    private void hasLocalLogging(Set<Integer> violationsSet, int variant, AbstractVertex<?> node) {
+        if (hasNodeCharacteristic(node, "Stereotype", "internal") && !hasNodeCharacteristic(node, "Stereotype", "local_logging")) {
+            violationsSet.add(10);
+            violationsSet.add(11);
+        }
+    }
+
+    // The local logging agent should sanitize the log data and remove any PII, passwords, API keys, etc.
+    private void hasLogSanitization(Set<Integer> violationsSet, int variant, AbstractVertex<?> node) {
+        if (hasNodeCharacteristic(node, "Stereotype", "local_logging") && !hasNodeCharacteristic(node, "Stereotype", "log_sanitization")) {
+            violationsSet.add(11);
+        }
+    }
+
+    // A message broker should be used to realize the communication between local logging agent and central logging
+    // system. These two should use mutual authentication and encrypt all transmitted data and availability should
+    // be ensured by providing periodic health and status data.
+    private void hasMessageBroker(Set<Integer> violationsSet, int variant, AbstractVertex<?> node) {
+        if (hasNodeCharacteristic(node, "Stereotype", "logging_server")
+                && !hasDataCharacteristicAcrossVariables(node, "Stereotype", "message_broker")) {
+            violationsSet.add(12);
+        }
+    }
+
+    // Secrets should be managed centrally following the Secret as a Service principle.
+    private void hasSecretManager(AbstractTransposeFlowGraph aTFG, Map<Integer, List<AbstractTransposeFlowGraph>> existenceViolations) {
+        existenceViolations.putIfAbsent(18, new ArrayList<AbstractTransposeFlowGraph>());
+        if (!hasNodeWithCharacteristic(aTFG, "Stereotype", "secret_manager")) {
+            existenceViolations.get(18)
+                    .add(aTFG);
+        }
+    }
+
+    private boolean hasNodeCharacteristic(AbstractVertex<?> node, String type, String value) {
+        return node.getAllVertexCharacteristics()
+                .stream()
+                .anyMatch(n -> n.getTypeName()
+                        .equals(type)
+                        && n.getValueName()
+                                .equals(value));
+    }
+
+    private boolean hasDataCharacteristicAcrossVariables(AbstractVertex<?> node, String type, String value) {
+        return node.getAllDataCharacteristics()
+                .stream()
+                .anyMatch(v -> v.getAllCharacteristics()
+                        .stream()
+                        .anyMatch(c -> c.getTypeName()
+                                .equals(type)
+                                && c.getValueName()
+                                        .equals(value)));
+    }
```

Comment:
Are these still used?
This functionality is provided by the `AbstractVertex` Class

Reply:
yes, i did not see a method that checks whether a characteristic is an element. But the query could be optimized with the functions, but i am not sure about that

Reply:
Rename to `hasDataCharacteristicInAnyVariable` and add javadoc